master/internal/stream/trials.go

```diff
@@ -24,43 +25,47 @@ type TrialMsg struct {
 	bun.BaseModel `bun:"table:trials"`
 
 	// immutable attributes
-	ID int                      `bun:"id,pk" json:"id"`
-	TaskID model.TaskID         `bun:"task_id" json:"task_id"`
-	ExperimentID int            `bun:"experiment_id" json:"experiment_id"`
-	RequestID *model.RequestID  `bun:"request_id" json:"request_id"`
-	Seed int64                  `bun:"seed" json:"seed"`
-	HParams JsonB               `bun:"hparams" json:"hparams"`
+	ID           int              `bun:"id,pk" json:"id"`
+	TaskID       model.TaskID     `bun:"task_id" json:"task_id"`
+	ExperimentID int              `bun:"experiment_id" json:"experiment_id"`
+	RequestID    *model.RequestID `bun:"request_id" json:"request_id"`
+	Seed         int64            `bun:"seed" json:"seed"`
+	HParams      JsonB            `bun:"hparams" json:"hparams"`
 
 	// warmstart checkpoint id?
 
 	// mutable attributes
-	State model.State           `bun:"state" json:"state"`
-	StartTime time.Time         `bun:"start_time" json:"start_time"`
-	EndTime *time.Time          `bun:"end_time" json:"end_time"`
-	RunnerState string          `bun:"runner_state" json:"runner_state"`
-	Restarts int                `bun:"restarts" json:"restarts"`
-	Tags JsonB                  `bun:"tags" json:"tags"`
+	State       model.State `bun:"state" json:"state"`
+	StartTime   time.Time   `bun:"start_time" json:"start_time"`
+	EndTime     *time.Time  `bun:"end_time" json:"end_time"`
+	RunnerState string      `bun:"runner_state" json:"runner_state"`
+	Restarts    int         `bun:"restarts" json:"restarts"`
+	Tags        JsonB       `bun:"tags" json:"tags"`
 
 	// metadata
-	Seq int64                   `bun:"seq" json:"seq"`
+	Seq int64 `bun:"seq" json:"seq"`
 
-	// total batches?
+	// permission scope
+	WorkspaceID int `bun:"workspace_id" json:"workspace_id"`
```

Comment:
question: what are the pros and cons of passing WorkspaceID to the end user, as part of the TrialMsg?  Is that a good idea or a bad idea?

Reply:
Is this a conversation around exporting/not exporting the `WorkspaceID` field or about including it in `TrialMsg` all together?

Pro: 
- Opens up the possibility of subscribing to trials at a workspace level (not implemented)
- Allows us to have a single data model for relevant Trial data for streaming, rather than a separate one containing workspace_id values
    - Alternatively, we could choose not export `WorkspaceID` and only make it accessible inside the `stream` package
- Allows us to include the workspace_id in the `TrialMsg` events so we don't makes additional calls to the DB
- Also avoids synchronization errors that could arise from reactively querying for the workspace_id in the event that the workspace_id has changed in between the `Event` being added into the notify queue & being the message being broadcasted
 
Con:
- Unclear how likely workspace-level trial subscriptions are to be used (most likely a moot point)
- Exposes workspace_id values to end user (not reflective of the workspace name) and could cause undo confusion

Reply:
This is a question about the user interface.  That's usually what I think of first.  I think when you say "export" you mean to other packages, but I'm talking about "exposing" it in terms of "serializing it to streaming clients".

In this case, "user interface" a little murky because I don't want to expose streaming updates to end users.  So in this case, "user interface" is more about "what interface does backend expose to other teams".

If WorkspaceID is not part of the normal user interface, because it's an id and end users think in terms of names, then I'd say that's a darn good reason to not expose it to end users.  (I think `json:"-"` accomplishes that).

Also, in general, I'd like for the promise of the Streaming Updates project to be "we stream rows of a table".  It keeps things simple to document and simple to understand.  One of the big mistakes we made with our proto definitions is that our proto definitions include all sorts of joined information from all sorts of tables.  It makes the proto api difficult to implement and dangerous to consume.  For instance, you might want to list all experiments over the REST API but the query will be much more expensive than that because it's going to also join a half-dozen other tables across the whole database.

So for that reason as well, I'd prefer we not expose it.

I agree with your various pros about reducing race conditions and avoiding additional queries.  I think having WorkspaceID be a field on the TrialMsg is fine and good because it makes life a lot simpler inside the go code.  And I like simple.

Lastly:

> Opens up the possibility of subscribing to trials at a workspace level (not implemented)

It's always much easier to extend an interface than shrink it, so I'd much prefer we expose WorkspaceID to streaming clients on the day we decide we want to allow subscribing to trials at the workspace level.  Which I don't think we should do because I don't think we should be offering filtering based on joined information.  RBAC forces us to do this particular join, so it's not totally out of the question.  But let's be conservative about what we expose, and expose more as we learn that we need to.

Reply:
I agree, I like the idea of keeping `WorkspaceID`, and excluding it from the json. I'll go with this for the time being.

___
One thing you did bring up as far as the desire to make sure the "we stream rows of a table" is that in order to do permissions/RBAC, as it currently stands, this is based knowing what workspace a trial belongs to, which is something that a trial doesn't currently know about itself without joining to other tables.

Alternatively we can move the scope down to the experiment level which the trial knows about; however, then the creation of a new experiment would result in a subscription accessScopeSet being stale and require reconnect with our current design.

Another approach could be hiding the join behind `trials_augmented_view`; however, since it's not a materialized view it won't be cached. So there wouldn't be any performance benefits.

Reply:
> One thing you did bring up as far as the desire to make sure the "we stream rows of a table" is that in order to do permissions/RBAC, as it currently stands, this is based knowing what workspace a trial belongs to, which is something that a trial doesn't currently know about itself without joining to other tables

Let's avoid `trials_augmented_view`, since that is a relic of our huge protobuf objects, and is exactly the sort of thing we need to avoid, for performance reasons.

I know there will be some joins that happen in order to implement RBAC.  As long as we don't send WorkspaceID as part of the TrialMsg to the streaming clients, it can remain purely an implementation detail, and we can change how it's implemented later without having to consult with different teams that might be consuming the streaming APIs.

I don't have the database design chops to confidently say if that's better, or if we should start copying WorkspaceID around to every object to make streaming easier, but in the absence of better schema chops I'd prefer we not change the schema, if we can get away with that.